### ISSUE TITLE
Trouble installing and getting mod_wsgi working (Fatal Python error: init_fs_encoding: failed to get the Python codec...)

### ISSUE THREAD

**Issue by Shark-waters** [GrahamDumpleton/mod_wsgi#869]
Hi Graham,

Firstly, I really appreciate the work you have done/are doing on this mod_wsgi project. I am new to web app development and have been doing a lot of reading including many of your guides and issue responses. 

I am trying to deploy a django project from a working development application to production. In development its working on a Win10 machine, Python3.8 and using the normal `python manage.py runserver` commands from django 4.2.7. I'm using a virtual environment via CMD command line statements.
```
pip install virtualenv
virtualenv venv
venv\scripts\activate
```

Moving to my QA server (basically a Win10 machine again: Python3.8 system installation, virtualenv with Python 3.8 and same site-packages as in development) 

I have then:

- Installed Apache24 : httpd-2.4.58-win64-VS17 from [Apache Lounge](https://www.apachelounge.com/download/)
- I put it into *C:\Apache\Apache* following most of this really good[ guide](https://medium.com/@deepankaracharyya/deploying-a-django-app-with-apache-mod-wsgi-windows-e9712dc07bba) by Deepankar
- I updated my windows environment variables to add `MOD_WSGI_APACHE_ROOTDIR = "C:\Apache\Apache"`. (Note: changes need a reboot to become effective).  I did read your comment that backslashes should not be used but only forward slashes. However my windows system would not work with forward slashes and all other path type variables use backslash. 
- Installed Apache with
 ```
httpd.exe -k install
```
- I also installed [Microsoft Visual Studio Build tools](https://visualstudio.microsoft.com/downloads/?q=build+tools) vs 2022 but I am not sure that I needed those? As I then install mod_wsgi via pip by

```
cd C:\Users\myproj
venv\scripts\activate
pip install mod_wsgi
```

Did I need C++ build tools doing it this non CMMI way? I was following a lot of the guidance from the excellent [mod_wsgi project description](https://pypi.org/project/mod-wsgi/)

Starting the apache server with `httpd.exe -k start` and navigating to localhost yields the "It works" web page from apache. 
Then moving to configuring the httpd.conf file to reference my django app and mod_wsgi I get into trouble. I removed django from the equation and configured for a simple wsgi app called testApp.py (quoted further down). My httpd.conf looks like

```
# My additions for testApp.py
################################

Define SRVROOT "C:/Apache24/Apache24"
serverName localhost
ServerRoot "${SRVROOT}"

# From Tutorial
# From running "mod_wsgi-express module-config"
LoadFile "c:/users/prxpdhil/appdata/local/programs/python/python38/python38.dll"
LoadModule wsgi_module "C:/Users/prxpdhil/django-projects/testing-wsgi/venv/lib/site-packages/mod_wsgi/server/mod_wsgi.cp38-win_amd64.pyd"
WSGIPythonHome "C:/Users/prxpdhil/django-projects/testing-wsgi/venv"

# From Django deploy documentation
WSGIScriptAlias / C:/Users/prxpdhil/django-projects/testing-wsgi/testApp.py
WSGIPythonPath C:/Users/prxpdhil/django-projects/testing-wsgi

<Directory C:/Users/prxpdhil/django-projects/testing-wsgi>
    Require all granted
</Directory>

# End of my additions
#######################################
```

Note: After running `pip install mod_wsgi`  it compiles and produces the file *mod_wsgi.cp38-win_amd64.pyd* but running `mod_wsgi-express module-config' says the file should be called just *mod_wsgi.pyd*. I am wondering if this is part of my problem i.e. its compiled for a different environment than the one I actually have? I have tried renaming the file and also, as above, just changing the reference. 

Running 'hhtpd.exe -k restart' after updating the above conf file to point to the testApp.py then gives apache error log:

> Starting the 'Apache2.4' service
The 'Apache2.4' service is running.
pm_winnt:notice] [pid 23020:tid 480] AH00455: Apache/2.4.58 (Win64) mod_wsgi/5.0.0 Python/3.8 configured -- resuming normal operations
[Thu Nov 30 10:32:41.029293 2023] [mpm_winnt:notice] [pid 23020:tid 480] AH00456: Apache Lounge VS17 Server built: Oct 18 2023 13:03:18
[Thu Nov 30 10:32:41.030291 2023] [core:notice] [pid 23020:tid 480] AH00094: Command line: 'C:\\Apache24\\Apache24\\bin\\httpd.exe -d C:/Apache24/Apache24'
[Thu Nov 30 10:32:41.037257 2023] [mpm_winnt:notice] [pid 23020:tid 480] AH00418: Parent: Created child process 3928
Python path configuration:
  PYTHONHOME = (not set)
  PYTHONPATH = (not set)
  program name = 'python'
  isolated = 0
  environment = 1
  user site = 1
  import site = 1
  sys._base_executable = 'C:\\Apache24\\Apache24\\bin\\httpd.exe'
  sys.base_prefix = 'C:\\Users\\prxpdhil\\AppData\\Local\\Programs\\Python\\Python38'
  sys.base_exec_prefix = 'C:\\Users\\prxpdhil\\AppData\\Local\\Programs\\Python\\Python38'
  sys.executable = 'C:\\Apache24\\Apache24\\bin\\httpd.exe'
  sys.prefix = 'C:\\Users\\prxpdhil\\AppData\\Local\\Programs\\Python\\Python38'
  sys.exec_prefix = 'C:\\Users\\prxpdhil\\AppData\\Local\\Programs\\Python\\Python38'
  sys.path = [
    'C:\\Users\\prxpdhil\\AppData\\Local\\Programs\\Python\\Python38\\python38.zip',
    '.\\DLLs',
    '.\\lib',
    'C:\\Apache24\\Apache24\\bin',
  ]
Fatal Python error: init_fs_encoding: failed to get the Python codec of the filesystem encoding
Python runtime state: core initialized
ModuleNotFoundError: No module named 'encodings'
>
>Current thread 0x00001150 (most recent call first):
><no Python frame>
>[Thu Nov 30 10:32:41.815826 2023] [mpm_winnt:crit] [pid 23020:tid 480] AH00419: master_main: create child process failed. Exiting.

I tried adding in PYTHONPATH and PYTHONHOME environment variables (advice from various people is not to do this, but I tried anyway to help debug) but that did not work either. 

My testApp.py file:

```
def application(environ, start_response):
    status = '200 OK'
    output = b'Hello World!'

    response_headers = [('Content-type', 'text/plain'),
                        ('Content-Length', str(len(output)))]
    start_response(status, response_headers)

    return [output]
```

I have a feeling my issue is a misconfiguration that I have made or wrong installation process as the error suggests python is just not found in some way. Another thing I've observed is that I see a Python36.dll in another venv folder that I have. Even although I  updated that to be python3.8. Might be red herring as I am not pointing to that or loading that. I've checked permissions on the various folders and the admin account I am working with and SYSTEM have access. 

Any guidance appreciated. 
Thanks. Alasdair

**Comment by GrahamDumpleton** [comment]
It looks like you installed Python only for the current user. Because Apache often runs as a special Windows user, Python needs to be installed for all users on the Windows machine, not just your own user. This is to ensure that certain environment variables which Python uses to find where it is installed are available to all users, including the special Windows user Apache runs as.

That is only suggest can offer right now. Getting Apache with mod_wsgi can be a pain to get working on Windows because Python broke the ability to embed Python on Windows in other applications and so hacks are used in mod_wsgi to configure Python so it might work. For reasons not understood, sometimes it still doesn't work and don't know what it is about some peoples environments which are different.

**Comment by Shark-waters** [comment]
Thanks Graham. 

I took your advice and cleaned up the Python installations on the machine. Indeed clearing out an older version of Python lurking, my user only version and then reinstalling with a *custom install* of Python ticking the "Install for all users" option did the trick. 

My simple test app is now working  i.e. being served by Apache and mod_wsgi. My main site is serving, just some issues with serving the static files to resolve. 

It does seem that wsgi on windows is much trickier than unix/linux. I noticed that uwsgi needed some hacks on the code to try to get this working on windows. The other Django deployment options for wsgi and asgi don't support windows e.g. gunicorn, Daphne. So for me mod_wsgi is the main option for windows. 
Cheers, Alasdair
